### ISSUE TITLE
Discussion: Easy Inventory with Barcodes, QR-codes, micro-codes.

### ISSUE THREAD

**Issue by a7hybnj2** [inventree/InvenTree#708]
I know this isn't the right place for a discussion but lacking a forum or IRC channel I thought I would try here.

I am interested in speeding up my inventory management by using bar codes both existing and adding my own. I know there is support for generating QR codes from inventree but some parts from mouser or digikey already come with micro-codes printed on them. Also, my PCB supplier often prints micro codes if I have margins on my boards.

What do you people use? Is there a USB code scanner you people like? What kind of barcode printer do you use to add bar codes? Is there a way to input the feed from a webcam to read the codes?

@SchrodingersGat Is there a way to add an existing bar code as an alias to a part? For instance if I decode the micro-code that comes on a bag of parts and it doesn't align with existing part information could it be added and searched for as an alias to a part?

Thanks

**Comment by SchrodingersGat** [comment]
@a7hybnj2 see the discussion over at https://github.com/inventree/InvenTree/issues/677

**Comment by SchrodingersGat** [comment]
> What do you people use? Is there a USB code scanner you people like? What kind of barcode printer do you use to add bar codes? Is there a way to input the feed from a webcam to read the codes?

In the other thread there's discussion of a javascript library for reading codes via a webcam. I want to look into that, it looks quite promising.

> @SchrodingersGat Is there a way to add an existing bar code as an alias to a part? For instance if I decode the micro-code that comes on a bag of parts and it doesn't align with existing part information could it be added and searched for as an alias to a part?

Not yet but it's a great idea. I think it should be developed in the near future. Using some sort of "ID" field separate to the internal stock-id for doing lookup would be easy to support.

> What kind of barcode printer do you use to add bar codes

I do not currently have this directly integrated with InvenTree, I print a "picture" of the QR code generated in InvenTree web interface. What I would like to achieve is basically native print support for label printers so you can just press a "print" button on the QR code in the web interface.

**Comment by a7hybnj2** [comment]
Thanks! I look forward to what comes next.

I also think bar codes for stock location would be good. For example my feeders and my PnP machine appear to have UUID barcodes that were put on by the manufacturer. Would be nice to associate that with a part or as a stock location.

**Comment by SchrodingersGat** [comment]
Fair point. Currently InvenTree supports internal barcodes for StockLocation objects, having support for some other key would be good too.

**Comment by SchrodingersGat** [comment]
Do you currently use a barcode scanner or printer?

**Comment by a7hybnj2** [comment]
No. At my last office we had a zebra printer that you could use different sticker sizes to print codes with info. I haven’t invested in my own yet.

I am in the market for a wireless code scanner but haven’t settled on one. Qr code and data matrix support seems pretty common but micro qr seems harder to find.

I would like to buy one in the next few days if I can track one down.

**Comment by rco3** [comment]
Having spent quite a bit of bandwidth on this problem over the last few days, I'm leaning more and more in the direction that QR codes from InvenTree are the way to go here. 

I've developed a scheme for matching arrays of barcodes often found on part bags into templates; however, the deeper I dig, the more exceptions to the templates I'm finding.  The essential problem is the inability to know that ALL the barcodes in an image were captured, and that any given vendor varies their format too much to validate based on that.  I can see using these codes to inject data on a one-time basis, but for repeated stock tasks I believe that QR and DataMatrix formats are preferable due to their inherent error correction.

For handling Stock Items I believe that parts with a DataMatrix or QR code already present should use that as their UID for Stock tracking purposes.  For items like Digi-Key 1D barcodes, NAC or Newark with arrays of 1D codes, I believe that an InvenTree QR code should be attached to the item when captured into inventory. 

I/O:  I use a couple of relatively cheap, offshore handheld scanners for 1D and 2D barcodes. I don't think I paid more than $40 for either of them. The 2D scanner will also scan 1D barcodes, but its hard to be selective and a sea-of-barcodes-type label is pretty irritating to handle. The 1D scanner is more of a true line-scanner type, and actually scans where the red line is - plus is wireless.  A Bluetooth 2D handheld is in my immediate future.  A webcam and a few computer vision libraries make quick work of capturing all of these formats, but either streams a lot of video to the server to decode, or places a lot of burden on the user's system and environment. I think of this as an advanced user option.

As I've mentioned elsewhere, I use a Dymo Labelwriter 450.  Today, I copied the InvenTree QR code from the popup window, opened it in Preview (OS X), and printed directly to the 450.  I printed both to 4" x 2-1/8" and 2-1/8" x1-1/8" labels, and had no problem capturing either with my phone camera.  

In both sizes, there is significant room for additional content in human-readable form. Would be really nice to have a module that allowed Admin to design a form using various Stock Item fields for content that could be printed to the user's system - probably just need to know the current label size, but even sticking a half sheet of Letter or A4 in the bag with parts is really easy and effective. Even a simple default form with the QR code, Name, Description would be terrific. I could autogenerate that label from outside if the API provided access to the QR code for a stock item.

**Comment by SchrodingersGat** [comment]
@rco3 thanks for the feedback! A few things jump out at me:

### Bluetooth 2D barcode scanner

I do not have a barcode scanner (although it sounds like I should get one!) 

- Can they read data-matrix style codes?
- How does the data get to the computer? As a "keyboard" input device? I'd have to consider how to get that data into InvenTree

### Labels

I had not thought of printing "labels" but this is a really awesome idea. I like the idea of making this label printing "customizable". And, just recently I have started working on a framework that would make this pretty easy to do!!

InvenTree now supports a "plugin" interface - this is currently used as the scaffolding for the barcode system (which is under development). But I have also just started working towards a "report template" plugin - https://github.com/inventree/InvenTree/issues/740

This report template would allow the generation of custom PDF (or image) files which uses the built-in django templating language which has access to all the internal database models! So, it would be super straight forward to create a label-making plugin that users can customize simply by providing a new template file!

I will probably now make this label printing a front-line feature of the reporting plugin. It is such a good idea! If you want to assist or have some input on how you think this could/should work then please by all means feel free.

Cheers

**Comment by a7hybnj2** [comment]
Exciting! I hope adding barcodes improves the ease of inventory.

I just ordered https://www.amazon.com/dp/B07G5V95F9/ref=cm_sw_r_cp_api_i_v0SPEb351JN1E and will report back how well it works (should be here tmrw). I picked it because you can enable/disable each type of barcode and it is one of the few that I found that claims to handle micro-qr.

**Comment by rco3** [comment]
SchrodingersGat, 

I'm glad you liked the idea.

> ### Bluetooth 2D barcode scanner

Most 2D scanners read most of: PDF417, MicroPDF417, Data Matrix, Maxicode, QR Code, MicroQR, Aztec, Hanxin. My last (wired) one was $40.  I've got a $50 unit in the cart now that adds Bluetooth. Wired one would act as a keyboard (HID) or as a serial port.

> ### Labels

> 
> InvenTree now supports a "plugin" interface -
"report template" plugin - #740

I've been absorbing these changes. As I've been primarily a hardware guy for the last couple of decades, let me just say that I'm learning a great deal reading through your source.  It's a more sophisticated (Pythonic?) approach than I would have instinctively taken (hardware guy!), I'm looking forward to working with it. 

More specifically, I'm wondering if maybe this is the correct place to handle some sort of IPN auto-generator?

> This report template would allow the generation of custom PDF (or image) files which uses the built-in django templating language which has access to all the internal database models! So, it would be super straight forward to create a label-making plugin that users can customize simply by providing a new template file!
> 
Hand, meet glove. Perfect.

> I will probably now make this label printing a front-line feature of the reporting plugin. It is such a good idea! If you want to assist or have some input on how you think this could/should work then please by all means feel free.
> 

I'm glad to offer something of use. I would very much like to engage in a discussion about workflows and how to support them, both to understand your intent and concepts and to try to incorporate support for my own needs.

That said, I want to contribute usable code - but (hardware guy). I'm working through Django tutorials, and now that I've a working InvenTree install again (thanks!), I'm thinking I want to take a whack at fleshing out some Barcode plugin code...

                           ...speaking of which, I've got a list going of the fields in a DataMatrix, and it looks like DK and Mouser are using a lot of the same fields and prefixes for them; without querying either vendor's API, we can extract 

10D: Date code
K: (Cust PO?)
1K: SO# (Sales Order?)
10K: Invoice #
11K: MSL (? - DK),  Invoice # (Mouser)
4L: Country of Origin
P: SKU / Cust P/N (I believe this is 'Part', and stuffed with SKU if the customer doesn't provide one)
1P: MFG P/N
Q: Quantity
1T: Lot code
1V: Manufacturer
11Z: ?
12Z: Part ID
13Z: LOAD ID
20Z: zero padding

if they are present.  Not all codes are present in all DataMatrix.  The above comes from matching DM field content to bag labels. 

Arrow uses a smaller DataMatrix with a single field (no delimiters), but I'm not yet sure which parts are which.  They also have an API; haven't gotten that far. There may be others I don't yet have analysis of; there are two standards, EIA 706 and EIA 802 that I don't yet have access to but I believe to address the standards used by the vendors.  I may just purchase them, they aren't that expensive.

Given some subset of the above data, can we use this to directly capture new Stock Items into the inventory? Or to simplify/pre-populate for new entry? We can definitely use them as UID-type keys, in place of printing our own labels, but I honestly would like the option to use InvenTree QR codes as well.

Look, I'm sorry for heaving a big wall-o-text every time I comment.  It's a topic that simply CONSUMES me right now and I'm afraid it shows. I love what you're doing here. I'd like to engage in a more in-depth discussion of workflow and InvenTree structure, and I don't want to verbally crap all over your repo, or try to monopolize all your bandwidth. Is this the right venue? Email? Smoke signals? Shut up?



(some of my parts have 2D barcodes on them.  I haven't read any yet. I bet there's serial numbers... Do InvenTree QRs have serials?)(SHUT UP!)

**Comment by SchrodingersGat** [comment]
> I've been absorbing these changes. As I've been primarily a hardware guy for the last couple of decades, let me just say that I'm learning a great deal reading through your source.

I'm an embedded hardware guy myself, so welcome aboard! 

> Given some subset of the above data, can we use this to directly capture new Stock Items into the inventory? Or to simplify/pre-populate for new entry? We can definitely use them as UID-type keys, in place of printing our own labels, but I honestly would like the option to use InvenTree QR codes as well.

This data would absolutely be useful for capturing new stock items into the system. As much data as can be pre-populated, the better! As for the "UID" a simple hash of the barcode data should be *enough* to uniquely use the manufacturers barcode for stock-lookup. But there's nothing stopping you from printing out the InvenTree barcode too!

With regard to the Inventree barcode, I've been thinking about improving upon that by having a default "Inventree Label" which includes some human readable data alongside the existing QR. This label could be adjusted using a template, of course

**Comment by SchrodingersGat** [comment]
support for printing of custom labels is now implemented - https://github.com/inventree/InvenTree/pull/906

**Comment by matmair** [comment]
@SchrodingersGat we support all of this now - I think. Should we close this?

**Comment by SchrodingersGat** [comment]
This has been hanging around for a while! Yes the barcode support is much more mature now, I think this can definitely be closed out.
